### PR TITLE
[SPARK-18621][PYTHON] Make sql type reprs eval-able

### PR DESCRIPTION
**Pull request by crflynn**
### What changes were proposed in this pull request?

These changes update the `__repr__` methods of type classes in `pyspark.sql.types` to print string representations which are `eval`-able. In other words, any instance of a `DataType` will produce a repr which can be passed to `eval()` to create an identical instance.

Similar changes previously submitted: https://github.com/apache/spark/pull/25495

### Why are the changes needed?

This [bug](https://issues.apache.org/jira/browse/SPARK-18621) has been around for a while. The current implementation returns a string representation which is valid in scala rather than python. These changes fix the repr to be valid with python.

The [motivation](https://docs.python.org/3/library/functions.html#repr) is "to return a string that would yield an object with the same value when passed to eval()".

### Does this PR introduce _any_ user-facing change?

Example:

Current implementation:

```python
from pyspark.sql.types import *

struct = StructType([StructField('f1', StringType(), True)])
repr(struct)
# StructType(List(StructField(f1,StringType,true)))
new_struct = eval(repr(struct))
# Traceback (most recent call last):
#   File "<input>", line 1, in <module>
#   File "<string>", line 1, in <module>
# NameError: name 'List' is not defined

struct_field = StructField('f1', StringType(), True)
repr(struct_field)
# StructField(f1,StringType,true)
new_struct_field = eval(repr(struct_field))
# Traceback (most recent call last):
#   File "<input>", line 1, in <module>
#   File "<string>", line 1, in <module>
# NameError: name 'f1' is not defined
```

With changes:

```python
from pyspark.sql.types import *

struct = StructType([StructField('f1', StringType(), True)])
repr(struct)
# StructType([StructField('f1', StringType(), True)])
new_struct = eval(repr(struct))
struct == new_struct
# True

struct_field = StructField('f1', StringType(), True)
repr(struct_field)
# StructField('f1', StringType(), True)
new_struct_field = eval(repr(struct_field))
struct_field == new_struct_field
# True
```

### How was this patch tested?

The changes include a test which asserts that an instance of each type is equal to the `eval` of its `repr`, as in the above example.

### CODE REVIEW

**Review by srowen — COMMENTED** [review]
Seems reasonable to me. I suppose this is a user-facing change and always some concern over breaking something, but, seems like the intent of `__repr__` can only be to be evaluatable, right?

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by crflynn** [comment]
I fixed the related dataframe test. I'm not sure how the k8s failures are related to these changes.

**Comment by crflynn** [comment]
I think I've got everything passing. There were a lot of doctests that needed to be updated.

**Comment by crflynn** [comment]
Updated once more to resolve conflicts.

**Comment by zero323** [comment]
In general it looks reasonable (it is common to want to reuse inferred schema for example) ‒ my only concern is that it is going to break user snapshot tests and possibly affect things like language-to-language migrations. So it ifs going to be merged it has to be included as potentially breaking change in the release notes.

**Comment by crflynn** [comment]
Should we include a note in the `migration_guide` under pyspark docs?

**Comment by crflynn** [comment]
Just wanted to check-in on the PR status here. If there is still some risk of merging I could always release this as a separate package which patches in the reprs.

**Comment by zero323** [comment]
cc @HyukjinKwon

**Review by srowen — COMMENTED** [review]
Yeah I don't know how to evaluate the risk of breaks. I think it's reasonable. I agree that a quick note in the migration guide would be safe

File: python/pyspark/pandas/spark/utils.py [hunk 1]
```diff
@@ -143,9 +150,15 @@ def force_decimal_precision_scale(
     ...             StructField('b', StringType(), True)])),
     ...     StructField("B", DecimalType(30, 15), False)]),
     ...     precision=30, scale=15)  # doctest: +NORMALIZE_WHITESPACE
-    StructType(List(StructField(A,StructType(List(StructField(a,MapType(DecimalType(30,15),\
-ArrayType(DecimalType(30,15),false),false),false),StructField(b,StringType,true))),true),\
-StructField(B,DecimalType(30,15),false)))
+    StructType([StructField('A',
+                            StructType([StructField('a',
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/spark/utils.py:154]: Cab we get rid of all these white spaces? or at least two space indentation.

**Comment by HyukjinKwon** [comment]
I kind of like this change .. and can't think of a case this change breaks something for now .. so I am fine with this. By right, yeah __repr__  should return something `eval`uable but that's that the case in many projects outside in practice though.

File: python/pyspark/pandas/spark/utils.py (hunk 1 above)
Comment (crflynn) [reply python/pyspark/pandas/spark/utils.py:154]: I'll take another look. IIRC there was some nuance with doctests and wrapped whitespace that was difficult to work around, which is why it has the odd hanging indents here.

**Comment by HyukjinKwon** [comment]
@crflynn mind running `dev/reformat-python` script to reformat the codes?

**Review by HyukjinKwon — APPROVED** [review]
Otherwise, LGTM from me side. It follows what `__repr__` is supposed to be according to the official Python docs. cc @BryanCutler @viirya @ueshin FYI

**Review by BryanCutler — APPROVED** [review]
LGTM

**Review by viirya — APPROVED** [review]

**Review by viirya — COMMENTED** [review]
Looks reasonable. It's nice to add a note into the migration guide.

**Review by srowen — APPROVED** [review]

**Pull request closed without merging**